### PR TITLE
 Refactors species-based offsets, condenses all offsets into a single list, adds support for all other standing icons

### PR DESCRIPTION
**Pull request by Militaires**
this means you can offset every individual shoe, hat, ID, uniform and whatever.

### CODE REVIEW

File: code/modules/mob/living/carbon/human/species.dm [hunk 1]
```diff
@@ -399,7 +398,10 @@
 				else
 					hair_overlay.color = forced_colour
 				hair_overlay.alpha = hair_alpha
-				hair_overlay.pixel_y += hair_y_offset
+				if("face" in H.dna.species.offset_features)
+					var/list/offset_list = H.dna.species.offset_features["face"]
+					hair_overlay.pixel_x += offset_list[1]
+					hair_overlay.pixel_y += offset_list[2]
```
Comment (Jordie0608) [inline code/modules/mob/living/carbon/human/species.dm]: No need for another list in any of these, just do
`hair_overlay.pixel_y += H.dna.species.offset_features["face"][2]`

Comment (Militaires) [reply code/modules/mob/living/carbon/human/species.dm]: thanks

**Comment by Iamgoofball** [comment]
:+1: will help with dwarves

File: code/modules/mob/living/carbon/human/species.dm [hunk 2]
```diff
@@ -18,8 +18,7 @@
 
 	var/sexes = 1		// whether or not the race has sexual characteristics. at the moment this is only 0 for skeletons and shadows
 
-	var/face_y_offset = 0
-	var/hair_y_offset = 0
+	var/offset_features = list("uniform" = list(0,0), "id" = list(0,0), "gloves" = list(0,0), "glasses" = list(0,0), "ears" = list(0,0), "shoes" = list(0,0), "s_store" = list(0,0), "mask" = list(0,0), "head" = list(0,0),"face" = list(0,0), "belt" = list(0,0), "back" = list(0,0), "suit" = list(0,0), "neck" = list(0,0))
```
Comment (AnturK) [inline code/modules/mob/living/carbon/human/species.dm]: Don't use magic strings.

File: code/modules/mob/living/carbon/update_icons.dm [hunk 3]
```diff
@@ -125,6 +125,13 @@
 	if(wear_mask)
 		if(!(head && (head.flags_inv & HIDEMASK)))
 			overlays_standing[FACEMASK_LAYER] = wear_mask.build_worn_icon(state = wear_mask.icon_state, default_layer = FACEMASK_LAYER, default_icon_file = 'icons/mob/mask.dmi')
+			if(ishuman(src))
```
Comment (AnturK) [inline code/modules/mob/living/carbon/update_icons.dm]: No. If it's human functionality you do it on human level.

File: code/modules/mob/living/carbon/human/species.dm (hunk 2 above)
Comment (Militaires) [reply code/modules/mob/living/carbon/human/species.dm]: does that mean i should add it to the species on init or what

File: code/modules/mob/living/carbon/update_icons.dm (hunk 3 above)
Comment (Militaires) [reply code/modules/mob/living/carbon/update_icons.dm]: sure thing boss

File: code/modules/mob/living/carbon/human/species.dm [hunk 4]
```diff
@@ -8,6 +8,23 @@
...
+#define OFFSET_EARS "ears"
+#define OFFSET_SHOES "shoes"
+#define OFFSET_S_STORE "s_store"
+#define OFFSET_FACEMASK "mask"
+#define OFFSET_HEAD "head"
+#define OFFSET_FACE "face"
+#define OFFSET_BELT "belt"
+#define OFFSET_BACK "back"
+#define OFFSET_SUIT "suit"
+#define OFFSET_NECK "neck"
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/species.dm]: These need to be under the __DEFINES directory

File: code/modules/mob/living/carbon/human/species.dm [hunk 5]
```diff
@@ -399,7 +415,9 @@
 				else
 					hair_overlay.color = forced_colour
 				hair_overlay.alpha = hair_alpha
-				hair_overlay.pixel_y += hair_y_offset
+				if("face" in H.dna.species.offset_features)
+					hair_overlay.pixel_x += H.dna.species.offset_features["face"][1]
+					hair_overlay.pixel_y += H.dna.species.offset_features["face"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/species.dm]: Just making the defines isn't enough, you have to actualy *use* the things.

File: code/modules/mob/living/carbon/human/species.dm [hunk 6]
```diff
@@ -420,14 +438,18 @@
 		if(H.lip_style && (LIPS in species_traits) && HD)
 			var/mutable_appearance/lip_overlay = mutable_appearance('icons/mob/human_face.dmi', "lips_[H.lip_style]", -BODY_LAYER)
 			lip_overlay.color = H.lip_color
-			lip_overlay.pixel_y += face_y_offset
+			if("face" in H.dna.species.offset_features)
+				lip_overlay.pixel_x += H.dna.species.offset_features["face"][1]
+				lip_overlay.pixel_y += H.dna.species.offset_features["face"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/species.dm]: Same here

File: code/modules/mob/living/carbon/human/species.dm [hunk 7]
```diff
@@ -420,14 +438,18 @@
...
 			standing += lip_overlay
 
 		// eyes
 		if((EYECOLOR in species_traits) && HD)
 			var/mutable_appearance/eye_overlay = mutable_appearance('icons/mob/human_face.dmi', "eyes", -BODY_LAYER)
 			eye_overlay.color = "#" + H.eye_color
-			eye_overlay.pixel_y += face_y_offset
+			if("face" in H.dna.species.offset_features)
+				eye_overlay.pixel_x += H.dna.species.offset_features["face"][1]
+				eye_overlay.pixel_y += H.dna.species.offset_features["face"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/species.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 8]
```diff
@@ -134,6 +134,10 @@ There are several things that need to be remembered:
 		if(!uniform_overlay)
 			uniform_overlay = U.build_worn_icon(state = "[t_color]", default_layer = UNIFORM_LAYER, default_icon_file = 'icons/mob/uniform.dmi', isinhands = FALSE)
 
+
+		if("uniform" in dna.species.offset_features)
+			uniform_overlay.pixel_x += dna.species.offset_features["uniform"][1]
+			uniform_overlay.pixel_y += dna.species.offset_features["uniform"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 9]
```diff
@@ -147,14 +151,20 @@ There are several things that need to be remembered:
...
 		if(client && hud_used && hud_used.hud_shown)
 			client.screen += wear_id
 		update_observer_view(wear_id)
 
 		//TODO: add an icon file for ID slot stuff, so it's less snowflakey
-		overlays_standing[ID_LAYER] = wear_id.build_worn_icon(state = wear_id.item_state, default_layer = ID_LAYER, default_icon_file = 'icons/mob/mob.dmi')
+		id_overlay = wear_id.build_worn_icon(state = wear_id.item_state, default_layer = ID_LAYER, default_icon_file = 'icons/mob/mob.dmi')
+		if("id" in dna.species.offset_features)
+			id_overlay.pixel_x += dna.species.offset_features["id"][1]
+			id_overlay.pixel_y += dna.species.offset_features["id"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 10]
```diff
@@ -186,7 +197,11 @@ There are several things that need to be remembered:
 		if(!t_state)
 			t_state = gloves.icon_state
 		overlays_standing[GLOVES_LAYER] = gloves.build_worn_icon(state = t_state, default_layer = GLOVES_LAYER, default_icon_file = 'icons/mob/hands.dmi')
-
+		gloves_overlay = overlays_standing[GLOVES_LAYER]
+		if("gloves" in dna.species.offset_features)
+			gloves_overlay.pixel_x += dna.species.offset_features["gloves"][1]
+			gloves_overlay.pixel_y += dna.species.offset_features["gloves"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 11]
```diff
@@ -209,7 +224,11 @@ There are several things that need to be remembered:
 		if(!(head && (head.flags_inv & HIDEEYES)) && !(wear_mask && (wear_mask.flags_inv & HIDEEYES)))
 
 			overlays_standing[GLASSES_LAYER] = glasses.build_worn_icon(state = glasses.icon_state, default_layer = GLASSES_LAYER, default_icon_file = 'icons/mob/eyes.dmi')
-
+		var/mutable_appearance/glasses_overlay = overlays_standing[GLASSES_LAYER]
+		if("glasses" in dna.species.offset_features)
+			glasses_overlay.pixel_x += dna.species.offset_features["glasses"][1]
+			glasses_overlay.pixel_y += dna.species.offset_features["glasses"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 12]
```diff
@@ -231,7 +250,11 @@ There are several things that need to be remembered:
 		update_observer_view(ears,1)
 
 		overlays_standing[EARS_LAYER] = ears.build_worn_icon(state = ears.icon_state, default_layer = EARS_LAYER, default_icon_file = 'icons/mob/ears.dmi')
-
+		var/mutable_appearance/ears_overlay = overlays_standing[EARS_LAYER]
+		if("ears" in dna.species.offset_features)
+			ears_overlay.pixel_x += dna.species.offset_features["ears"][1]
+			ears_overlay.pixel_y += dna.species.offset_features["ears"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 13]
```diff
@@ -252,6 +275,11 @@ There are several things that need to be remembered:
 				client.screen += shoes					//add it to client's screen
 		update_observer_view(shoes,1)
 		overlays_standing[SHOES_LAYER] = shoes.build_worn_icon(state = shoes.icon_state, default_layer = SHOES_LAYER, default_icon_file = 'icons/mob/feet.dmi')
+		var/mutable_appearance/shoes_overlay = overlays_standing[SHOES_LAYER]
+		if("shoes" in dna.species.offset_features)
+			shoes_overlay.pixel_x += dna.species.offset_features["shoes"][1]
+			shoes_overlay.pixel_y += dna.species.offset_features["shoes"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 14]
```diff
@@ -272,13 +300,25 @@ There are several things that need to be remembered:
 		if(!t_state)
 			t_state = s_store.icon_state
 		overlays_standing[SUIT_STORE_LAYER]	= mutable_appearance('icons/mob/belt_mirror.dmi', t_state, -SUIT_STORE_LAYER)
-
+		var/mutable_appearance/s_store_overlay = overlays_standing[SUIT_LAYER]
+		if("s_store" in dna.species.offset_features)
+			s_store_overlay.pixel_x += dna.species.offset_features["s_store"][1]
+			s_store_overlay.pixel_y += dna.species.offset_features["s_store"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 15]
```diff
@@ -272,13 +300,25 @@ There are several things that need to be remembered:
...
 
 /mob/living/carbon/human/update_inv_head()
 	..()
 	update_mutant_bodyparts()
+	if(head)
+		var/mutable_appearance/head_overlay = overlays_standing[HEAD_LAYER]
+		remove_overlay(HEAD_LAYER)
+		if("head" in dna.species.offset_features)
+			head_overlay.pixel_x += dna.species.offset_features["head"][1]
+			head_overlay.pixel_y += dna.species.offset_features["head"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 16]
```diff
@@ -298,7 +338,11 @@ There are several things that need to be remembered:
 			t_state = belt.icon_state
 
 		overlays_standing[BELT_LAYER] = belt.build_worn_icon(state = t_state, default_layer = BELT_LAYER, default_icon_file = 'icons/mob/belt.dmi')
-
+		var/mutable_appearance/belt_overlay = overlays_standing[BELT_LAYER]
+		if("belt" in dna.species.offset_features)
+			belt_overlay.pixel_x += dna.species.offset_features["belt"][1]
+			belt_overlay.pixel_y += dna.species.offset_features["belt"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 17]
```diff
@@ -319,7 +363,11 @@ There are several things that need to be remembered:
 		update_observer_view(wear_suit,1)
 
 		overlays_standing[SUIT_LAYER] = wear_suit.build_worn_icon(state = wear_suit.icon_state, default_layer = SUIT_LAYER, default_icon_file = 'icons/mob/suit.dmi')
-
+		var/mutable_appearance/suit_overlay = overlays_standing[SUIT_LAYER]
+		if("suit" in dna.species.offset_features)
+			suit_overlay.pixel_x += dna.species.offset_features["suit"][1]
+			suit_overlay.pixel_y += dna.species.offset_features["suit"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 18]
```diff
@@ -351,8 +399,24 @@ There are several things that need to be remembered:
 
 /mob/living/carbon/human/update_inv_wear_mask()
 	..()
+	var/mutable_appearance/mask_overlay = overlays_standing[FACEMASK_LAYER]
+	remove_overlay(FACEMASK_LAYER)
+	if("mask" in dna.species.offset_features)
+		mask_overlay.pixel_x += dna.species.offset_features["mask"][1]
+		mask_overlay.pixel_y += dna.species.offset_features["mask"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 19]
```diff
@@ -351,8 +399,24 @@ There are several things that need to be remembered:
...
+	apply_overlay(FACEMASK_LAYER)
 	update_mutant_bodyparts() //e.g. upgate needed because mask now hides lizard snout
 
+/mob/living/carbon/human/update_inv_back()
+	..()
+	var/mutable_appearance/back_overlay = overlays_standing[BACK_LAYER]
+	remove_overlay(BACK_LAYER)
+	if("back" in dna.species.offset_features)
+		back_overlay.pixel_x += dna.species.offset_features["back"][1]
+		back_overlay.pixel_y += dna.species.offset_features["back"][2]
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/update_icons.dm]: Same here

File: code/modules/mob/living/carbon/human/species.dm (hunk 4 above)
Comment (duncathan) [reply code/modules/mob/living/carbon/human/species.dm]: if they're only used in this file he can leave them here and then `#undef` them at the end of the file

Comment (ninjanomnom) [reply code/modules/mob/living/carbon/human/species.dm]: They're not only used here, they're also used in update_icons.dm

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 20]
```diff
@@ -124,7 +124,7 @@ There are several things that need to be remembered:
 		else if(U.adjusted == DIGITIGRADE_STYLE)
 			t_color = "[t_color]_l"
 
-		var/mutable_appearance/uniform_overlay
+		var/mutable_appearance/uniform_overlay = overlays_standing[UNIFORM_LAYER]
```
Comment (AnturK) [inline code/modules/mob/living/carbon/human/update_icons.dm]: You changed the default here so the uniform won't update for species without gender variants.

File: code/modules/mob/living/carbon/human/species.dm [hunk 21]
```diff
@@ -18,8 +17,7 @@
 
 	var/sexes = 1		// whether or not the race has sexual characteristics. at the moment this is only 0 for skeletons and shadows
 
-	var/face_y_offset = 0
-	var/hair_y_offset = 0
+	var/offset_features = list(OFFSET_UNIFORM = list(0,0), OFFSET_ID = list(0,0), OFFSET_GLOVES = list(0,0), OFFSET_GLASSES = list(0,0), OFFSET_EARS = list(0,0), OFFSET_SHOES = list(0,0), OFFSET_S_STORE = list(0,0), OFFSET_FACEMASK = list(0,0), OFFSET_HEAD = list(0,0), OFFSET_FACE = list(0,0), OFFSET_BELT = list(0,0), OFFSET_BACK = list(0,0), OFFSET_SUIT = list(0,0), OFFSET_NECK = list(0,0))
```
Comment (lzimann) [inline code/modules/mob/living/carbon/human/species.dm]: `var/list/offset_features = ...`

Comment (Militaires) [reply code/modules/mob/living/carbon/human/species.dm]: how embarrassing

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 22]
```diff
@@ -351,8 +399,24 @@ There are several things that need to be remembered:
 
 /mob/living/carbon/human/update_inv_wear_mask()
 	..()
+	var/mutable_appearance/mask_overlay = overlays_standing[FACEMASK_LAYER]
+	remove_overlay(FACEMASK_LAYER)
+	if(OFFSET_FACEMASK in dna.species.offset_features)
```
Comment (AnturK) [inline code/modules/mob/living/carbon/human/update_icons.dm]: You need to check if overlay actually exists here in and in the one below

**Pull request merged**